vpr/src/place/net_cost_handler.h

```diff
@@ -25,135 +28,442 @@ enum class e_cost_methods {
+     */
+    /* [0...cluster_ctx.clb_nlist.nets().size()-1] -> 3D bounding box*/
+    vtr::vector<ClusterNetId, t_bb> ts_bb_coord_new_, ts_bb_edge_new_;
+    /* [0...cluster_ctx.clb_nlist.nets().size()-1][0...num_layers-1] -> 2D bonding box on a layer*/
+    vtr::vector<ClusterNetId, std::vector<t_2D_bb>> layer_ts_bb_coord_new_, layer_ts_bb_edge_new_;
+    /* [0...cluster_ctx.clb_nlist.nets().size()-1][0...num_layers-1] -> number of sink pins on a layer*/
+    vtr::Matrix<int> ts_layer_sink_pin_count_;
+    /* [0...num_affected_nets] -> net_id of the affected nets */
+    std::vector<ClusterNetId> ts_nets_to_update_;
 
-/**
- * @brief Free (layer_)ts_bb_edge_new, (layer_)ts_bb_coord_new, ts_layer_sink_pin_count, and ts_nets_to_update data structures.
- */
-void free_try_swap_net_cost_structs();
 
-void set_net_handlers_placer_state(PlacerState& placer_state);
+    /**
+     * @brief In each of these vectors, there is one entry per cluster level net:
+     * [0...cluster_ctx.clb_nlist.nets().size()-1].
+     * net_cost and proposed_net_cost: Cost of a net, and a temporary cost of a net used during move assessment.
+     * We also use negative cost values in proposed_net_cost as a flag to indicate that
+     * the cost of a net has not yet been updated.
+     * bb_update_status: Flag array to indicate whether the specific bounding box has been updated
+     * in this particular swap or not. If it has been updated before, the code
+     * must use the updated data, instead of the out-of-date data passed into the
+     * subroutine, particularly used in try_swap(). The value NOT_UPDATED_YET
+     * indicates that the net has not been updated before, UPDATED_ONCE indicated
+     * that the net has been updated once, if it is going to be updated again, the
+     * values from the previous update must be used. GOT_FROM_SCRATCH is only
+     * applicable for nets larger than SMALL_NETS and it indicates that the
+     * particular bounding box is not incrementally updated, and hence the
+     * bounding box is got from scratch, so the bounding box would definitely be
+     * right, DO NOT update again.
+     */
+    vtr::vector<ClusterNetId, double> net_cost_;
+    vtr::vector<ClusterNetId, double> proposed_net_cost_;
+    vtr::vector<ClusterNetId, NetUpdateState> bb_update_status_;
+
+    /**
+     * @brief Matrices below are used to precompute the inverse of the average
+     * number of tracks per channel between [subhigh] and [sublow].  Access
+     * them as chan?_place_cost_fac[subhigh][sublow].  They are used to
+     * speed up the computation of the cost function that takes the length
+     * of the net bounding box in each dimension, divided by the average
+     * number of tracks in that direction; for other cost functions they
+     * will never be used.
+     */
+    vtr::NdMatrix<float, 2> chanx_place_cost_fac_; // [0...device_ctx.grid.width()-2]
+    vtr::NdMatrix<float, 2> chany_place_cost_fac_; // [0...device_ctx.grid.height()-2]
```

Comment:
TODO: make these line up with the edge of the grid so we get rid of some if statements (probably width-1 or width instead of width-2).
Also, we should have a 3rd dimension normalization somewhere using the same technique and it seems that variable should also go here.